File: 백준/Gold/17298. 오큰수/오큰수.py
import sys

input = sys.stdin.readline

N = int(input()) # 4
A = list(map(int,input().split())) # 3 5 2 7
# 오큰수 NGE(i) 오른쪽에 있으면서 Ai보다 큰 수 중에 가장 왼쪽에 있는 수 3의 오큰수는 5 / 5의 오큰수는 7 / 2의 오큰수는 7 / 7의 오큰수는 -1

answer = [-1] * N
stack = [0]

for i in range(1, N):
    while stack:
        if A[stack[-1]] < A[i]:
            answer[stack.pop()] = A[i]
        else:
            break
    stack.append(i)

print(*answer)

File: 백준/Gold/17298. 오큰수/오큰수.js
const input = require('fs').readFileSync(process.version === 'v18.17.1' ? 'input.txt' : '/dev/stdin')
  .toString().trim().split("\n")

const solution = () => {
  const N = +input[0];
  const numbers = input[1].split(" ").map(Number);

  let stack = [];
  let result = new Array(N).fill(-1);  // 처음에 모든 결과를 -1로 초기화

  for (let i = 0; i < N; i++) {
    while (stack.length !== 0 && numbers[stack[stack.length - 1]] < numbers[i]) {
      result[stack.pop()] = numbers[i];
    }
    stack.push(i);
  }

  console.log(result.join(' '));
};

solution();